test_util/helpers.py

```diff
@@ -23,81 +24,137 @@ def path_join(p1, p2):
+            self.netloc,
+            self.path,
+            self.query if self.query else '',
+            self.fragment if self.fragment else ''))
+
+    def get_url(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """return new Url with any component replaced
+        """
+        return Url(
+            scheme if scheme is not None else self.scheme,
+            host if host is not None else self.host,
+            path if path is not None else self.path,
+            query if query is not None else self.query,
+            fragment if fragment is not None else self.fragment,
+            port if port is not None else self.port)
+
 
-    def __init__(self, default_host_url, api_base, default_headers=None,
-                 ca_cert_path=None, get_node_url=None):
-        self.default_host_url = default_host_url
-        self.api_base = api_base
-        if default_headers is None:
-            default_headers = dict()
-        self.default_headers = default_headers
-        self.ca_cert_path = ca_cert_path
-        self._get_node_url = get_node_url
-
-    def api_request(self, method, path, host_url=None, node=None, port=None, **kwargs):
+class ApiClient:
+    """utilizes requests.session with some URL handling to remove boilerplate for making
+    cluster requests. Can also be used to bind advanced helper methods of a given service
+    """
+    def __init__(self, url: Url, get_node_port=None):
         """
-        Makes a request with default headers + user auth headers if necessary
-        If self.ca_cert_path is set, this method will pass it to requests
         Args:
-            method: (str) name of method for requests.request
-            path: see get_url()
-            host_url: override the client's default host URL
-            node: if a get_node_url function is set, node and port will be passed to this function
-                instead of setting a host_url or using the default
-            port: can be used with node with get_node_url is set
-            **kwargs: any optional arguments to be passed to requests.request
+            url: Url object to which requests will be made
+        Keyword Args:
+            get_node_port: a callback that takes a node string as an argument. This function
+                must return the port for that string. Intended for communicating with an API
+                that uses different ports on different hosts (e.g. AdminRouter, Mesos )
         """
-        headers = copy.copy(self.default_headers)
+        self.url = url
+        self.session = requests.Session()
+        self._get_node_port = get_node_port
 
-        # allow kwarg to override verification so client can be used generically
-        if self.ca_cert_path and 'verify' not in kwargs:
-            kwargs['verify'] = self.ca_cert_path
+    def get_client(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """Takes the same arguments as Url.get_url
```

Comment:
Why not `get_client(self, url)` with `url` being an object of type `Url`? Then there is only one interface to maintain.

In an unrelated note, it would be great of the docstring could say  more about the purpose of this method. It looks like the fundamental idea is to be able to re-use existing ApiClient objects, but mutate their `url` attribute. Is that the core idea? In that case I don't quite like the name `clone`.

Just a proposal, but what do you think about having an _actual_ `.clone()` method that does what it says and then consider `.url` (or `default_url`) as public interface?

In _using_ code, this could look like this (I find this rather expressive):

```
# Generate an actual clone of this client.
new_client = apiclient.clone()

# Override its url prefix:
new_client.default_url = Url(...)
```

Of course one could also do `new_client = apiclient.clone(new_default_url=...)`, which I also would find expressive enough.